optimisation/design/Third_Party_API_Design.md

```diff
@@ -0,0 +1,595 @@
+# An Application Programming Interface for user defined model functions in PACE
+**2020-05-15**
+
+
+# Scope
+
+This documents details the required function signatures for user defined sample scattering models
+and how they interface with Horace, using either Matlab, Python, C, C++ or Fortran.
+
+
+# Motivation
+
+Analysis of inelastic neutron scattering data almost always requires a parameterised model of the measured scattering intensities.
+Horace, as currently implemented, requires this model to be in the form of a Matlab function.
+This function is required to either 
+
+1. Take a 4D vector of momentum-energy ![\mathbb{Q} \equiv (\mathb{Q},E)] and yield the scattering function ![S(\mathbb{Q})].
+2. Take the 3D vector of momentum transfer ![\mathbf{Q}] only and yield ![N] sets of vectors ![(E_n, S(\mathbf{Q}, E_n))] where ![N] is the number of dispersion branches.
+
+For type 2, the ![N] sets of vectors are contained in a cell array
+and a helper function `disp2sqw` is used to obtain ![S(\mathbb{Q})]
+by convolving the calculated delta-function dispersion with broadening functions (Gaussian is default). 
+Whilst the internal fitting routines of Horace (including for resolution convolution) requires type 1 model functions (yielding ![S(\mathbb{Q})]),
+many actual user models are type 2, including `SpinW` and `Euphonic`.
+Thus, the new PACE implementation should allow both types of models.
+We also note, that when instrumental resolution convolution (`TobyFit`) is used,
+the broadening function in `disp2sqw` should only represent some intrinsic physical broadening
+(such as a quasiparticle lifetime).
+Whereas when `TobyFit` is not used,
+the broadening function in `disp2sqw` will be a combination of both intrinsic and instrumental resolution broadening.
+
+In the current Horace implementation, these model functions are used by the `multifit_sqw` function/class
+(which inherits from `multifit`) to fit to data using a built-in Levenberg-Marquardt optimizer. 
+The [new design for PACE](Model_Optimisation_Design.md) proposes instead to use a pair of classes, `OptFunction` and `OptModel`.
+`OptModel` will be more like the current `multifit` (or `tobyfit`) class,
+whilst `OptFunction` is a wrapper around the user supplied model function to handle parameter bindings and bounds currently handled by `multifit`.
+`OptFunction` should therefore have functionality to transform type 2 input to type 1 input
+whilst `OptModel` would expect only type 1 input.
```

Comment:
I agree that `OptFunction` must contain the ability to transform function **output** into the equivalent of type 1.
In the case of resolution convolution the class inheriting from `OptFunction` and implementing the Tobyfit method would do this conversion at the start of its scope, but other convolution methods might perform the energy-axis convolution analytically in which case type 2 (or my suggested type 3) function output would be converted to type 1 as part of the convolution.

Reply:
Ok. I've changed the text to note that for standard transformations listed in the optimisation document that `OptFunction` will automatically do the transformation of both input and output, and for others, the user is responsible for the transformation.